src/state.cc

```diff
@@ -92,12 +92,33 @@ Edge* State::AddEdge(const Rule* rule) {
   return edge;
 }
 
+Node* State::GetNode(const char* path, uint64_t slash_bits) {
+  return GetNode(StringPiece(path), slash_bits);
+}
+
 Node* State::GetNode(StringPiece path, uint64_t slash_bits) {
-  Node* node = LookupNode(path);
-  if (node)
-    return node;
-  node = new Node(path.AsString(), slash_bits);
-  paths_[node->path()] = node;
+  const std::pair<Paths::iterator, bool> result = paths_.try_emplace(path, nullptr);
+  Node*& node = result.first->second;
+  const bool isNew = result.second;
+  if (isNew) {
+    assert(node == nullptr);
+    node = new Node(path.AsString(), slash_bits);
+    // Update the key to refer to the long-lived string in the Node
```

Comment:
Oh this is smart but also a bit subtle. Can I ask you to add a comment explaining here that this cannot break the hash map's internals because the key type's equality and hash operations only care about the content of the StringPiece-pointed data, not the address it lives at. I think that would help for long-term maintenance.